### ISSUE TITLE
Socket disconnecting after a while

### ISSUE THREAD

**Issue by simonmcl** [WalletConnect/WalletConnectSwiftV2#739]
**Describe the bug**
Similar to https://github.com/WalletConnect/walletconnect-monorepo/issues/2019 i'm having an issue that after a few minutes of being connected the socket is disconnecting. After it disconnects I also get this issue: https://github.com/WalletConnect/WalletConnectSwiftV2/issues/685 where xcode complains about application state being checked on a background thread

**SDK Version**
WalletConnectSwiftV2 v1.4.1

**To Reproduce**
Setup as per instructions / sample app. All works fine as expected for first few minutes.

In case its relevant, I'm experimenting with using a wrapper around the native swift web sockets instead of starscream as I'm trying to keep a library as free of dependencies as possible, code is below. I tried to build some auto reconnect logic as i'm not sure if thats handled in the library or not, but that just got into a cycle of throwing errors and never successfully reconnected

```Swift
import Foundation
import WalletConnectRelay
import OSLog

public enum NativeSocketError: Error {
	case errorWithCode(URLSessionWebSocketTask.CloseCode)
}

public class NativeSocket: NSObject, WebSocketConnecting, URLSessionWebSocketDelegate {
	
	private var socket: URLSessionWebSocketTask? = nil
	
	init(withURL url: URL) {
		self.socket = URLSession.shared.webSocketTask(with: url)
		self.isConnected = false
		self.request = URLRequest(url: url)
		super.init()
		
		self.socket?.delegate = self
	}
	
	
	
	// MARK: - WebSocketConnecting
	
	public var isConnected: Bool
	
	public var onConnect: (() -> Void)?
	
	public var onDisconnect: ((Error?) -> Void)?
	
	public var onText: ((String) -> Void)?
	
	public var request: URLRequest
	
	public func connect() {
		socket?.resume()
	}
	
	public func disconnect() {
		socket?.cancel()
	}
	
	public func write(string: String, completion: (() -> Void)?) {
		let message = URLSessionWebSocketTask.Message.string(string)
		socket?.send(message) { err in
			if let e = err {
				os_log("NativeSocket sending error: %@", log: .default, type: .info, "\(e)")
				
			}
			
			if let comp = completion {
				comp()
			}
		}
	}
	
	
	
	// MARK: - URLSessionWebSocketDelegate
	
	public func urlSession(_ session: URLSession, webSocketTask: URLSessionWebSocketTask, didOpenWithProtocol protocol: String?) {
		isConnected = true
		os_log("NativeSocket connected", log: .default, type: .info)
		
		if let onC = onConnect {
			onC()
		}
		
		receiveMessage()
	}
	
	public func urlSession(_ session: URLSession, webSocketTask: URLSessionWebSocketTask, didCloseWith closeCode: URLSessionWebSocketTask.CloseCode, reason: Data?) {
		isConnected = false
		os_log("NativeSocket did close", log: .default, type: .error)
		
		if let onD = onDisconnect {
			if closeCode != URLSessionWebSocketTask.CloseCode.normalClosure {
				os_log("NativeSocket did close with code: normal", log: .default, type: .error)
				onD(NativeSocketError.errorWithCode(closeCode))
				
			} else {
				os_log("NativeSocket did close with code: %@", log: .default, type: .error, "\(closeCode)")
				onD(nil)
			}
		}
	}
	
	func receiveMessage() {
		socket?.receive(completionHandler: { [weak self] result in
			
			switch result {
				case .failure(let error):
					os_log("NativeSocket Error receiving: %@", log: .default, type: .error, "\(error)")
					
					// If its failing because the conneciton closed by itself, try to reconnect
					let nsErr = error as NSError
					if nsErr.code == 57 && nsErr.domain == "NSPOSIXErrorDomain" {
						self?.disconnect()
					}
					
				case .success(let message):
					switch message {
						case .string(let messageString):
							os_log("NativeSocket received message: %@", log: .default, type: .info, messageString)
							if let onT = self?.onText {
								onT(messageString)
							}
							
						case .data(let data):
							os_log("NativeSocket received data: %@", log: .default, type: .info, data.description)
							if let onT = self?.onText {
								onT(data.description)
							}
							
						default:
							os_log("NativeSocket received unknown data", log: .default, type: .info)
					}
			}
			
			if self?.isConnected == true {
				self?.receiveMessage()
			}
		})
	}
}

struct NativeSocketFactory: WebSocketFactory {
	
	func create(with url: URL) -> WebSocketConnecting {
		return NativeSocket(withURL: url)
	}
}
```

**Device (please complete the following information):**
 - Simulators and release devices running iOS 16

**Comment by llbartekll** [comment]
hey @simonmcl 

>after a few minutes of being connected the socket is disconnecting.

does it happen with Starscream or your implementation of URLSession?

**Comment by simonmcl** [comment]
@llbartekll using URLSession

**Comment by llbartekll** [comment]
@simonmcl we just released fix for https://github.com/WalletConnect/WalletConnectSwiftV2/issues/685
could you check if issue still exists in 1.5.1?

**Comment by llbartekll** [comment]
>but that just got into a cycle of throwing errors and never successfully reconnected

btw. could you share the logs

**Comment by simonmcl** [comment]
@llbartekll Yes 1.5.1 fixes the background state issue.

Re: logs
All I have is the error domain:
```
Error Domain=NSPOSIXErrorDomain Code=57 "Socket is not connected"
```
Which I believe is a timeout error that occurs around the 5 minute mark, although i'm not sure if its on the client side or the server side. I tried to catch that error and call `WebSocketConnecting.disconnect()` followed by `WebSocketConnecting.connect()` and it just goes around in a loop printing this error. I'm assuming there needs to be some sort of an inbuilt keep alive message passed on the connection every minute or so in order to keep the connection up

**Comment by simonmcl** [comment]
Had some time to play around with this. Yes its disconnecting exactly every 5 minutes. Its a timeout error, but I can't see a way to increase the timeout. WC would need to send some sort of a keepalive every minute or so to ensure sockets remain open across all types of apps. As a workaround for the minute, I recreate the socket + reconnect every time I get this error and it seems to keep the socket up

**Comment by ajay-simform** [comment]
Hello @simonmcl,
I came across the same issue you mentioned earlier regarding the socket disconnecting after 5 minutes. If you have any workarounds, could you kindly assist me by sharing them?

Thank you.

**Comment by simonmcl** [comment]
@ajay-simform yes I mentioned my workaround in the last sentence of my previous message

**Comment by llbartekll** [comment]
hey @ajay-simform, what sdk version do you use, the issue was supposed to be fixed in one of the recent releases.

**Comment by llbartekll** [comment]
here:
https://github.com/WalletConnect/WalletConnectSwiftV2/releases/tag/1.5.12
but please better upgrade to the latest

**Comment by llbartekll** [comment]
thanks for letting me know

**Comment by llbartekll** [comment]
@ajay-simform do you use starscream or urlsession?

**Comment by simonmcl** [comment]
@llbartekll updated to the latest `1.5.14` and can confirm nothings changed. Using URLSession it still dies after 5 minutes

**Comment by ajay-simform** [comment]
@llbartekll I am using URLSession as well

**Comment by simonmcl** [comment]
@alexander-lsvk yes listening to changes on the request property worked. 

Heres my full code for anyone that can't/doesn't want to use starscream

```Swift
import Foundation
import WalletConnectRelay
import OSLog

public enum NativeSocketError: Error {
	case errorWithCode(URLSessionWebSocketTask.CloseCode)
}

public class NativeSocket: NSObject, WebSocketConnecting, URLSessionWebSocketDelegate {
	
	private var socket: URLSessionWebSocketTask? = nil
	private var initURL: URL
	
	public init(withURL url: URL) {
		self.initURL = url
		self.isConnected = false
		self.request = URLRequest(url: url)
		super.init()
	}
	
	public func reconnect() {
		self.socket = URLSession.shared.webSocketTask(with: initURL)
		self.socket?.delegate = self
		self.isConnected = false
		self.connect()
	}
	
	
	
	// MARK: - WebSocketConnecting
	
	public var isConnected: Bool
	
	public var onConnect: (() -> Void)?
	
	public var onDisconnect: ((Error?) -> Void)?
	
	public var onText: ((String) -> Void)?
	
	public var request: URLRequest {
		didSet {
			os_log("NativeSocket new request set", log: .default, type: .info)
			
			if let url = request.url {
				self.socket = URLSession.shared.webSocketTask(with: url)
				self.socket?.delegate = self
				self.isConnected = false
				self.initURL = url
			}
		}
	}
	
	public func connect() {
		os_log("NativeSocket connect func called", log: .default, type: .info)
		socket?.resume()
	}
	
	public func disconnect() {
		os_log("NativeSocket disconnect func called", log: .default, type: .info)
		socket?.cancel()
	}
	
	public func write(string: String, completion: (() -> Void)?) {
		let message = URLSessionWebSocketTask.Message.string(string)
		socket?.send(message) { err in
			if let e = err {
				os_log("NativeSocket sending error: %@", log: .default, type: .info, "\(e)")
				
			}
			
			if let comp = completion {
				comp()
			}
		}
	}
	
	
	
	// MARK: - URLSessionWebSocketDelegate
	
	public func urlSession(_ session: URLSession, webSocketTask: URLSessionWebSocketTask, didOpenWithProtocol protocol: String?) {
		isConnected = true
		os_log("NativeSocket connected", log: .default, type: .info)
		
		if let onC = onConnect {
			onC()
		}
		
		receiveMessage()
	}
	
	public func urlSession(_ session: URLSession, webSocketTask: URLSessionWebSocketTask, didCloseWith closeCode: URLSessionWebSocketTask.CloseCode, reason: Data?) {
		isConnected = false
		os_log("NativeSocket did close", log: .default, type: .error)
		
		if let onD = onDisconnect {
			if closeCode != URLSessionWebSocketTask.CloseCode.normalClosure {
				os_log("NativeSocket did close with code: normal", log: .default, type: .error)
				onD(NativeSocketError.errorWithCode(closeCode))
				
			} else {
				os_log("NativeSocket did close with code: %@", log: .default, type: .error, "\(closeCode)")
				onD(nil)
			}
		}
	}
	
	func receiveMessage() {
		socket?.receive(completionHandler: { [weak self] result in
			
			switch result {
				case .failure(let error):
					os_log("NativeSocket Error receiving: %@", log: .default, type: .error, "\(error)")
					
					// If its failing because the conneciton closed by itself, try to reconnect
					let nsErr = error as NSError
					if nsErr.code == 57 && nsErr.domain == "NSPOSIXErrorDomain" {
						self?.isConnected = false
					}
					
				case .success(let message):
					switch message {
						case .string(let messageString):
							os_log("NativeSocket received message: %@", log: .default, type: .info, messageString)
							if let onT = self?.onText {
								onT(messageString)
							}
							
						case .data(let data):
							os_log("NativeSocket received data: %@", log: .default, type: .info, data.description)
							if let onT = self?.onText {
								onT(data.description)
							}
							
						default:
							os_log("NativeSocket received unknown data", log: .default, type: .info)
					}
			}
			
			if self?.isConnected == true {
				self?.receiveMessage()
			}
		})
	}
}

struct NativeSocketFactory: WebSocketFactory {
	
	func create(with url: URL) -> WebSocketConnecting {
		return NativeSocket(withURL: url)
	}
}
```

**Comment by llbartekll** [comment]
Great! closing the issue then

**Comment by cao13646** [comment]
@available(iOS 15.0, *)
    open var delegate: URLSessionTaskDelegate?
not available on iOS14.0
